Sources/SignalRClient/HubConnection.swift

```diff
@@ -432,6 +441,56 @@ public class HubConnection {
     fileprivate func connectionDidReconnect() {
         initiateHandshake()
     }
+
+    private func resetKeepAlive() {
+        if connection.inherentKeepAlive {
+            return
+        }
+
+        logger.log(logLevel: .debug, message: "Reset keep alive")
+        keepAlivePingTask?.cancel()
+
+        keepAlivePingTask = DispatchWorkItem { self.sendKeepAlivePing() }
+
+        DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + keepAliveIntervalInSeconds) {
+            if let keepAlivePingTask = self.keepAlivePingTask {
+                keepAlivePingTask.perform()
+            }
+        }
+    }
+
+    private func sendKeepAlivePing() {
+        logger.log(logLevel: .debug, message: "Send keep alive called")
+        guard handshakeStatus.isHandled else {
+            logger.log(logLevel: .debug, message: "Send keep alive called but not connected")
+            keepAlivePingTask = nil
+            return
+        }
+
+        do {
+            let cachedPingMessage = try hubProtocol.writeMessage(message: PingMessage.instance)
+            logger.log(logLevel: .debug, message: "Send keep alive")
+            connection.send(data: cachedPingMessage, sendDidComplete: { error in
+                if let error = error {
+                    self.logger.log(logLevel: .error, message: "Keep alive Error recevied \(error.localizedDescription)")
+                    self.keepAlivePingTask = nil
+                } else {
+                    self.logger.log(logLevel: .debug, message: "Keep alive - Still alive")
+                    self.resetKeepAlive()
+                }
+            })
+        } catch {
+            // We don't care about the error. It should be seen elsewhere in the client.
+            // The connection is probably in a bad or closed state now, cleanup the timer so it stops triggering
+            logger.log(logLevel: .error, message: "Couldn't write keep alive message \(error.localizedDescription)")
+            keepAlivePingTask = nil
+        }
+    }
+
+    private func cleanUpKeepAlive() {
+        keepAlivePingTask?.cancel()
+        keepAlivePingTask = nil
```

Comment:
This seems race'y. If we were on line 452 and there was  a context switch we would run clean up but then we would create a new keepAlive which would end up running even though the connection was closed.

Reply:
I added some semaphores around the cleanup and the reset and replaced the nil setting with the cleanup, I hope it's less race'y now.